### PR TITLE
Incremental GC: Keep reachable previously expired GC commits

### LINKED ISSUE #5889: Incremental CGC: Keep expired yet reachable commits
The following scenario might happen in an incremental GC run:
Let's say that in run ID `A` commits `a` and `b` were expired.
Now a run with ID `B` is running and recognizes that `a` and `b` have expired previously, therefore GC should stop going through those commits’ paths and leave them out of the resulting CSV file. 
Let's also say that no other commit has expired in run `B`.
Great, now comes run `C`. Since commits `a` and `b` weren’t in run `B`’s CSV output file (they were left out because they were also in run `A`), they will be recalculated in the current run.

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
### Linked Issue

Examine it for more details:
Closes #5889

---

## Change Description

If previously expired commits (i.e. commits that expired on a previous GC run) have at least one active child commit, keep the previously expired commits in the commits CSV output for successive GC runs.

### CODE REVIEW

File: pkg/graveler/retention/expired_commits.go [hunk 1]
```diff
@@ -38,9 +38,9 @@ func GetGarbageCollectionCommits(ctx context.Context, startingPointIterator *GCS
 	// All commits reached are added to the active set, until and including the first commit performed before the start of the retention period.
 	// All further commits in the ancestry are added to the expired set. The iteration stops upon reaching a commit which exists in the previouslyExpired set, or the DAG root.
 	processed := make(map[graveler.CommitID]time.Time)
-	previouslyExpiredMap := make(map[graveler.CommitID]bool)
+	previouslyExpiredMap := make(map[graveler.CommitID]map[graveler.CommitID]struct{})
```
Comment (arielshaqed) [inline pkg/graveler/retention/expired_commits.go:41]: Can we document the two keys?

File: pkg/graveler/retention/expired_commits.go [hunk 2]
```diff
@@ -113,17 +114,25 @@ func GetGarbageCollectionCommits(ctx context.Context, startingPointIterator *GCS
...
 				// If the parent commit was expired in a previous GC run (incremental GC case) we can stop because it's
 				// both expired in this run and the previous (in this path)
 				if _, ok = previouslyExpiredMap[nextCommitID]; ok {
+					// Get the previously expired commit's list of known children and add the current commit to it.
+					var nextCommitChildren = previouslyExpiredMap[nextCommitID]
+					if _, ok := nextCommitChildren[currentCommitID]; !ok {
+						nextCommitChildren[currentCommitID] = struct{}{}
+						previouslyExpiredMap[nextCommitID] = nextCommitChildren
+					}
 					break
```
Comment (arielshaqed) [inline pkg/graveler/retention/expired_commits.go:128]:
```suggestion
				if nextCommitChildren, ok := previouslyExpiredMap[nextCommitID]; ok {
					// Add current commit to previously expired commit's list of known children.
					if _, ok := nextCommitChildren[currentCommitID]; !ok {
						nextCommitChildren[currentCommitID] = struct{}{}
						previouslyExpiredMap[nextCommitID] = nextCommitChildren
					}
					break
```
But TBH not sure what "previouslyExpiredMap" really holds.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Really need a comment around `previouslyExpiredMap` :-/

File: pkg/graveler/retention/expired_commits.go (hunk 2 above)
Comment (Jonathan-Rosenberg) [reply pkg/graveler/retention/expired_commits.go]: done

File: pkg/graveler/retention/expired_commits.go (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply pkg/graveler/retention/expired_commits.go]: done

File: pkg/graveler/retention/expired_commits.go [hunk 3]
```diff
@@ -39,9 +39,9 @@ func GetGarbageCollectionCommits(ctx context.Context, startingPointIterator *GCS
 	// All further commits in the ancestry are added to the expired set. The iteration stops upon reaching a commit which exists in the previouslyExpired set, or the DAG root.
 	processed := make(map[graveler.CommitID]time.Time)
 	// Mapping between previously expired commits to their direct children.
-	previouslyExpiredMap := make(map[graveler.CommitID]map[graveler.CommitID]struct{})
+	prevExpiredCommitsToChildrenMap := make(map[graveler.CommitID]map[graveler.CommitID]struct{})
```
Comment (arielshaqed) [inline pkg/graveler/retention/expired_commits.go:42]: Best var name ever!  Makes the "`// Mapping...`" comment that I asked for almost useless :-)

**Review by arielshaqed — APPROVED** [review]
StilLGTM! ✌🏽

**Pull request merged**